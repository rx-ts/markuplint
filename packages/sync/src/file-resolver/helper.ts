import path from 'path'

import { Config, ConfigSet, margeConfig } from '@markuplint/file-resolver'
import { cosmiconfigSync } from 'cosmiconfig'

import { tryRequirePkg } from '../helper'

import { loadConfigFile } from './load-config-file'

const explorer = cosmiconfigSync('markuplint')

export function search<T = Config>(dir: string, cacheClear: boolean) {
  if (!cacheClear) {
    explorer.clearCaches()
  }
  dir = path.dirname(dir)
  const result = explorer.search(dir)
  if (!result || result.isEmpty) {
    return null
  }
  return {
    filePath: result.filepath,
    config: result.config as T,
  }
}

export function load<T = Config>(filePath: string, cacheClear: boolean) {
  if (!cacheClear) {
    explorer.clearCaches()
  }
  const result = explorer.load(filePath)
  if (!result || result.isEmpty) {
    return null
  }
  return {
    filePath: result.filepath,
    config: result.config as T,
  }
}

// eslint-disable-next-line sonarjs/cognitive-complexity
export function recursiveLoad(
  config: Config,
  filePath: string,
  files: Set<string>,
  cacheClear: boolean,
): ConfigSet {
  const errs: Error[] = []
  const baseDir = path.dirname(filePath)
  if (config.extends) {
    const extendFiles = Array.isArray(config.extends)
      ? config.extends
      : [config.extends]
    for (const _file of extendFiles) {
      if (/^\.+\//.test(_file)) {
        const file = path.resolve(path.join(baseDir, _file))
        if (files.has(file)) {
          continue
        }
        const extendFileResult = loadConfigFile(file, true, cacheClear)
        if (!extendFileResult) {
          continue
        }
        files = new Set(files).add(file)
        config = margeConfig(extendFileResult.config, config)
      } else {
        try {
          const mod = tryRequirePkg<Config>(_file, true)
          files.add(_file)
          config = margeConfig(mod, config)
        } catch (err) {
          errs.push(err as Error)
        }
      }
    }
  }
  delete config.extends
  return {
    files,
    config,
    errs,
  }
}
