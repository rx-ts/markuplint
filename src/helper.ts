export function tryRequirePkg<T>(pkg: string): T | undefined
export function tryRequirePkg<T>(pkg: string, throwError: true): T
export function tryRequirePkg<T>(
  pkg: string,
  throwError?: boolean,
): T | undefined {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports, @typescript-eslint/no-var-requires
    const result = require(pkg) as T | { default: T }
    return 'default' in result ? result.default : result
  } catch (err) {
    if (throwError) {
      throw err
    }
  }
}
