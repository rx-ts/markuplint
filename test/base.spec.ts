import textlint from '@markuplint/rule-textlint'

import { loadRule } from './helper'

import { exec } from 'markuplint-sync'

describe('basic usage', () => {
  it('should just work', () => {
    expect(
      exec({
        sourceCodes: /* HTML */ `<html>
          <head>
            <title>Any Page</title>
          </head>
          <body>
            <h1>Any Page</h1>
            <p>Anonymous</p>
          </body>
        </html>`,
        config: {
          rules: {
            doctype: true,
          },
        },
        rules: [loadRule('doctype')!],
      })[0].results,
    ).toMatchSnapshot()

    expect(() =>
      exec({
        sourceCodes: 'content',
        config: {
          rules: {
            textlint: true,
          },
        },
        rules: [textlint],
      }),
    ).toThrow('`verifySync` finished async. Use `verify` instead')
  })
})
