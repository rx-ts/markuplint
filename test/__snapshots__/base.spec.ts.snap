// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic usage should just work 1`] = `
Array [
  Object {
    "col": 1,
    "line": 1,
    "message": "Required doctype",
    "raw": "",
    "ruleId": "doctype",
    "severity": "error",
  },
]
`;
